### ISSUE TITLE
'Stricter' rate limiter

### ISSUE THREAD

**Issue by guersam** [svroonland/rezilience#322]
I have a rate limit requirement for a third party API.

At first it sounded very easy with rezilience. But when I tested against the real server, several requests were blocked due to the rate limit. Here's a [minimal reproduction](https://gist.github.com/guersam/5ecccfb84c0d6c0f39f3182f91bacdc1):

```sh
$ scala-cli https://gist.github.com/guersam/5ecccfb84c0d6c0f39f3182f91bacdc1
```

### Expected Result
Given rate limit of 10 per second, the second 10 timestamps should be at least 1 second from start because the limit of 10 is already reached within the first second

### Actual Result
The 11th elements is emitted even before 0.3 seconds

### Related issue
https://github.com/zio/zio/issues/7888

**Comment by svroonland** [comment]
Hi @guersam, do you still experience this issue with the latest ZIO version?

**Comment by guersam** [comment]
@svroonland Yes, because [it's intended behavior](https://github.com/zio/zio/issues/7888#issuecomment-1450201542) of `throttleShapeZIO`.

I ended up with writing my own sliding window pipeline and it just works. If you think sliding window is more suitable than token bucket for rate limiters, I'm willing to contribute my implementation as well.

**Comment by svroonland** [comment]
@guersam I'd be interested in that implementation, definitely. Apologies for the late reply. Looking forward to a PR!
